python/cucim/tests/unit/clara/test_image_cache.py

```diff
@@ -0,0 +1,216 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+
+def test_get_nocache():
+    from cucim import CuImage
+    cache = CuImage.cache()
+
+    assert int(cache.type) == 0
+    assert cache.memory_size == 0
+    assert cache.memory_capacity == 0
+    assert cache.free_memory == 0
+    assert cache.size == 0
+    assert cache.capacity == 0
+    assert cache.hit_count == 0
+    assert cache.miss_count == 0
+
+    config = cache.config
+    # Check essential properties
+    #  {'type': 'nocache', 'memory_capacity': 1024, 'capacity': 5461,
+    #   'mutex_pool_capacity': 11117, 'list_padding': 10000,
+    #   'extra_shared_memory_size': 100, 'record_stat': False}
+    assert config['type'] == 'nocache'
+    assert not config['record_stat']
+
+
+def test_get_per_process_cache():
+    from cucim import CuImage
+    cache = CuImage.cache('per_process', memory_capacity=2048)
+    assert int(cache.type) == 1
+    assert cache.memory_size == 0
+    assert cache.memory_capacity == 2**20 * 2048
+    assert cache.free_memory == 2**20 * 2048
+    assert cache.size == 0
+    assert cache.capacity > 0
+    assert cache.hit_count == 0
+    assert cache.miss_count == 0
+
+    config = cache.config
+    # Check essential properties
+    #  {'type': 'per_process', 'memory_capacity': 2048, 'capacity': 10922,
+    #   'mutex_pool_capacity': 11117, 'list_padding': 10000,
+    #   'extra_shared_memory_size': 100, 'record_stat': False}
+    assert config['type'] == 'per_process'
+    assert config['memory_capacity'] == 2048
+    assert not config['record_stat']
+
+
+def test_get_shared_memory_cache():
+    from cucim import CuImage
+    cache = CuImage.cache('shared_memory', memory_capacity=2048)
+    assert int(cache.type) == 2
+    assert cache.memory_size == 0
+    # It allocates additional memory
+    assert cache.memory_capacity > 2**20 * 2048
+    assert cache.free_memory > 2**20 * 2048
+    assert cache.size == 0
+    assert cache.capacity > 0
+    assert cache.hit_count == 0
+    assert cache.miss_count == 0
+
+    config = cache.config
+    # Check essential properties
+    #  {'type': 'shared_memory', 'memory_capacity': 2048, 'capacity': 10922,
+    #   'mutex_pool_capacity': 11117, 'list_padding': 10000,
+    #   'extra_shared_memory_size': 100, 'record_stat': False}
+    assert config['type'] == 'shared_memory'
+    assert config['memory_capacity'] == 2048
+    assert not config['record_stat']
+
+
+def test_preferred_memory_capacity(testimg_tiff_stripe_32x24_16_jpeg):
+    from cucim import CuImage
+    from cucim.clara.cache import preferred_memory_capacity
+
+    img = CuImage(testimg_tiff_stripe_32x24_16_jpeg)
+
+    # same with `img.resolutions["level_dimensions"][0]`
+    image_size = img.size('XY')  # 32x24
+    tile_size = img.resolutions['level_tile_sizes'][0]  # 16x16
+    patch_size = (tile_size[0] * 2, tile_size[0] * 2)
```

Comment:
```suggestion
    patch_size = (tile_size[0] * 2, tile_size[1] * 2)
```

tile_size is a 2-tuple?